src/solver/optimisation/constraint_builder.h

```diff
@@ -0,0 +1,376 @@
+                                double coeff,
+                                int shift = 0,
+                                bool wrap = false,
+                                int delta = 0);
+
+    ConstraintBuilder& FinalStorage(unsigned int index,
+                                    double coeff,
+                                    int shift = 0,
+                                    bool wrap = false,
+                                    int delta = 0);
+
+    ConstraintBuilder& PositiveUnsuppliedEnergy(unsigned int index,
+                                                double coeff,
+                                                int shift = 0,
+                                                bool wrap = false,
+                                                int delta = 0);
+
+    ConstraintBuilder& NegativeUnsuppliedEnergy(unsigned int index,
+                                                double coeff,
+                                                int shift = 0,
+                                                bool wrap = false,
+                                                int delta = 0);
+    ConstraintBuilder& LayerStorage(unsigned area,
+                                    unsigned layer,
+                                    double coeff,
+                                    int shift = 0,
+                                    bool wrap = false,
+                                    int delta = 0);
+
+    class ConstraintBuilderInvalidOperator : public std::runtime_error
+    {
+    public:
+        ConstraintBuilderInvalidOperator(const std::string& error_message) :
+         std::runtime_error(error_message)
+        {
+        }
+    };
+
+    ConstraintBuilder& SetOperator(char op)
+    {
+        if (op == '<' || op == '=' || op == '>')
+        {
+            operator_ = op;
+        }
+        else
+            throw ConstraintBuilderInvalidOperator(std::string("Invalid operator: ") + op);
+
+        return *this;
+    }
+
+    ConstraintBuilder& equalTo()
+    {
+        operator_ = '=';
+        return *this;
+    }
+    ConstraintBuilder& lessThan()
+    {
+        operator_ = '<';
+        return *this;
+    }
+    ConstraintBuilder& greaterThan()
+    {
+        operator_ = '>';
+        return *this;
+    }
+
+    void build();
+    int NumberOfVariables() const
+    {
+        return nombreDeTermes_;
+    }
+
+private:
+    PROBLEME_HEBDO& problemeHebdo;
+    PROBLEME_ANTARES_A_RESOUDRE& problemeAResoudre;
+    const std::vector<CORRESPONDANCES_DES_VARIABLES>& varNative;
+
+    unsigned int hourInWeek_ = 0;
+
+    char operator_;
+    double rhs_ = 0;
+    int nombreDeTermes_ = 0;
+    // ConstraintNamer ConstraintNameManager;
+
+    int GetShiftedTimeStep(int shift, bool wrap, int delta) const;
+    void AddVariable(int index, double coeff);
+    Variable::VariableManager GetVariableManager( int shift = 0, bool wrap = false, int delta = 0);
+};
+
+class Constraint
```

Comment:
The name seems misleading: 
those classes actually don't represent individual constraints, but more constraint "types", or constraint factories.

For example, for the reader it's hard to understand why a constraint uses a "constraint builder":  one would expect a constraint builder to **create** constraints instead.
Renaming it would make more sense: a constraint type or factory uses a builder to actually create constraints.

Reply:
definitely!